### PR TITLE
mds: print each flag value

### PR DESCRIPTION
**Pull request by joscollin**
Fixes: https://tracker.ceph.com/issues/48683
Fixes: https://tracker.ceph.com/issues/48682

### CODE REVIEW

File: src/mds/MDSMap.cc [hunk 1]
```diff
@@ -222,7 +222,16 @@ void MDSMap::print(ostream& out) const
 {
   out << "fs_name\t" << fs_name << "\n";
   out << "epoch\t" << epoch << "\n";
-  out << "flags\t" << hex << flags << dec << "\n";
```
Comment (batrick) [inline src/mds/MDSMap.cc:225]: print the flags as hex too

File: src/include/ceph_fs.h [hunk 2]
```diff
@@ -272,14 +272,14 @@ struct ceph_mon_subscribe_ack {
...
-#define CEPH_MDSMAP_ALLOW_MULTIMDS_SNAPS	     (1<<4)  /* cluster alllowed to enable MULTIMDS
+#define CEPH_MDSMAP_NOT_JOINABLE                 1  /* standbys cannot join */
+#define CEPH_MDSMAP_DOWN                         CEPH_MDSMAP_NOT_JOINABLE /* backwards compat */
+#define CEPH_MDSMAP_ALLOW_SNAPS                  2  /* cluster allowed to create snapshots */
+/* deprecated #define CEPH_MDSMAP_ALLOW_MULTIMDS 4 cluster allowed to have >1 active MDS */
+/* deprecated #define CEPH_MDSMAP_ALLOW_DIRFRAGS 8 cluster allowed to fragment directories */
+#define CEPH_MDSMAP_ALLOW_MULTIMDS_SNAPS	     16  /* cluster alllowed to enable MULTIMDS
                                                             and SNAPS at the same time */
-#define CEPH_MDSMAP_ALLOW_STANDBY_REPLAY         (1<<5)  /* cluster alllowed to enable MULTIMDS */
+#define CEPH_MDSMAP_ALLOW_STANDBY_REPLAY         32  /* cluster alllowed to enable MULTIMDS */
```
Comment (batrick) [inline src/include/ceph_fs.h:282]: Undo these changes please. This is the more conventional way to define powers-of-2 flags since it's easier to read for humans and not susceptible to mistyping (e.g. 65336 instead of 65536).

File: src/mds/MDSMap.h [hunk 3]
```diff
@@ -552,6 +553,12 @@ class MDSMap {
   static bool state_transition_valid(DaemonState prev, DaemonState next);
 
   CompatSet compat;
+  std::map<int, std::string> flagmap = {
```
Comment (batrick) [inline src/mds/MDSMap.h:556]:
```suggestion
  inline static const std::map<int, std::string> flagmap = {
```

**Review by batrick — CHANGES_REQUESTED** [review]
also change `MDSMap::dump`.

File: src/mds/MDSMap.cc [hunk 4]
```diff
@@ -141,6 +141,10 @@ void MDSMap::dump(Formatter *f) const
 {
   f->dump_int("epoch", epoch);
   f->dump_unsigned("flags", flags);
+  f->dump_bool("not_joinable", not_joinable());
+  f->dump_bool("allow_snaps", allows_snaps());
+  f->dump_bool("allow_multimds_snaps", allows_multimds_snaps());
+  f->dump_bool("allow_standby_replay", allows_standby_replay());
```
Comment (joscollin) [inline src/mds/MDSMap.cc:147]: @batrick We could check which flags are present by iterating the flagmap and dump only those, alternatively

File: src/mds/FSMap.cc [hunk 5]
```diff
@@ -199,7 +199,7 @@ void FSMap::generate_test_instances(std::list<FSMap*>& ls)
 
 void FSMap::print(ostream& out) const
 {
-  out << "e" << epoch << std::endl;
+  out << "epoch: " << epoch << std::endl;
```
Comment (batrick) [inline src/mds/FSMap.cc:202]: This screws up a lot of greps we already use. Undo this too.

File: src/mds/MDSMap.cc (hunk 4 above)
Comment (batrick) [reply src/mds/MDSMap.cc:147]: This is fine but please put it in an object.

File: src/mon/MDSMonitor.cc [hunk 6]
```diff
@@ -1061,7 +1061,7 @@ bool MDSMonitor::preprocess_command(MonOpRequestRef op)
     }
 
     rdata.append(ds);
-    ss << "dumped fsmap epoch " << fsmapp->get_epoch();
+    ss << "dumped fsmap";
```
Comment (batrick) [inline src/mon/MDSMonitor.cc:1064]: undo this too

**Review by batrick — CHANGES_REQUESTED** [review]

**Comment by batrick** [comment]
Also add a test please.

**Comment by joscollin** [comment]
I'm rewriting the `lsflags` code for json output

File: src/mds/MDSMap.cc [hunk 7]
```diff
@@ -202,6 +203,18 @@ void MDSMap::dump(Formatter *f) const
   f->dump_int("standby_count_wanted", std::max(0, standby_count_wanted));
 }
 
+void MDSMap::dump_flags_state(Formatter *f) const
+{
+  if (f) {
```
Comment (batrick) [inline src/mds/MDSMap.cc:208]: don't think we need this test

**Review by batrick — COMMENTED** [review]
otherwise lgtm, just needs tests

File: qa/tasks/cephfs/test_misc.py [hunk 8]
```diff
@@ -198,6 +198,16 @@ def test_dump_inode_hexademical(self):
         info = self.fs.mds_asok(['dump', 'inode', hex(ino)])
         assert info['path'] == "/foo"
 
+    def test_fs_lsflags(self):
+        self.fs.set_joinable(False)
+        self.fs.set_allow_new_snaps(False)
+        self.fs.set_allow_multimds_snaps(False)
+        self.fs.set_allow_standby_replay(True)
+        lsflags = json.loads(self.fs.mon_manager.raw_cluster_cmd('fs', 'lsflags', self.fs.name, "--format=json-pretty"))
+        self.assertEqual(lsflags["not_joinable"], True)
```
Comment (joscollin) [inline qa/tasks/cephfs/test_misc.py:207]: @batrick `not_joinable` is displayed/used as `joinable` also. 
See the command `ceph fs set cephfs joinable`, which updates `not_joinable` value also. So why there are two representations for the same thing?

File: qa/tasks/cephfs/test_misc.py [hunk 9]
```diff
@@ -198,6 +198,16 @@ def test_dump_inode_hexademical(self):
...
+        self.assertEqual(lsflags["allow_snaps"], False)
+        #self.assertEqual(lsflags["allow_multimds_snaps"], False)
```
Comment (joscollin) [inline qa/tasks/cephfs/test_misc.py:209]: I'm not sure how to set `allow_multimds_snaps`. It doesn't work for me as mentioned in [MonCommands](https://github.com/ceph/ceph/blob/master/src/mon/MonCommands.h#L317). Any suggestions?

File: qa/tasks/cephfs/test_misc.py (hunk 8 above)
Comment (batrick) [reply qa/tasks/cephfs/test_misc.py:207]: Because it's confusing to have a command like:

`ceph fs set cephfs notjoinable X`

File: qa/tasks/cephfs/test_misc.py (hunk 9 above)
Comment (batrick) [reply qa/tasks/cephfs/test_misc.py:209]: What does the debug output show?

File: src/mds/MDSMap.cc [hunk 10]
```diff
@@ -202,6 +203,16 @@ void MDSMap::dump(Formatter *f) const
   f->dump_int("standby_count_wanted", std::max(0, standby_count_wanted));
 }
 
+void MDSMap::dump_flags_state(Formatter *f) const
+{
+    f->open_object_section("flags_state");
+    f->dump_bool("not_joinable", not_joinable());
```
[the remark this answers is not in the document]
Comment (batrick) [reply src/mds/MDSMap.cc:209]: and elsewhere

**Review by batrick — CHANGES_REQUESTED** [review]

File: qa/tasks/cephfs/test_misc.py (hunk 9 above)
Comment (joscollin) [reply qa/tasks/cephfs/test_misc.py:209]: I meant it cannot be set like [this](https://github.com/ceph/ceph/pull/39660/files#diff-501f5d864f91e98b1c214da58065308a41554b87ce6366f2314a84c19c287ca7R596). 

Gives the error:
```
2021-03-09 17:23:34,222.222 INFO:__main__:======================================================================
2021-03-09 17:23:34,222.222 INFO:__main__:ERROR: test_fs_lsflags (tasks.cephfs.test_misc.TestMisc)
2021-03-09 17:23:34,222.222 INFO:__main__:----------------------------------------------------------------------
2021-03-09 17:23:34,223.223 INFO:__main__:Traceback (most recent call last):
2021-03-09 17:23:34,223.223 INFO:__main__:  File "/home/jcollin/workspace/ceph/qa/tasks/cephfs/test_misc.py", line 204, in test_fs_lsflags
2021-03-09 17:23:34,223.223 INFO:__main__:    self.fs.set_allow_multimds_snaps(False)
2021-03-09 17:23:34,223.223 INFO:__main__:  File "/home/jcollin/workspace/ceph/qa/tasks/cephfs/filesystem.py", line 596, in set_allow_multimds_snaps
2021-03-09 17:23:34,223.223 INFO:__main__:    self.set_var("allow_multimds_snaps", yes)
2021-03-09 17:23:34,223.223 INFO:__main__:  File "/home/jcollin/workspace/ceph/qa/tasks/cephfs/filesystem.py", line 575, in set_var
2021-03-09 17:23:34,223.223 INFO:__main__:    self.mon_manager.raw_cluster_cmd("fs", "set", self.name, var, *a)
2021-03-09 17:23:34,223.223 INFO:__main__:  File "/home/jcollin/workspace/ceph/build/../qa/tasks/vstart_runner.py", line 1005, in raw_cluster_cmd
2021-03-09 17:23:34,223.223 INFO:__main__:    return self.run_cluster_cmd(**kwargs).stdout.getvalue()
2021-03-09 17:23:34,223.223 INFO:__main__:  File "/home/jcollin/workspace/ceph/build/../qa/tasks/vstart_runner.py", line 995, in run_cluster_cmd
2021-03-09 17:23:34,223.223 INFO:__main__:    return self.controller.run(**kwargs)
2021-03-09 17:23:34,223.223 INFO:__main__:  File "/home/jcollin/workspace/ceph/build/../qa/tasks/vstart_runner.py", line 399, in run
2021-03-09 17:23:34,223.223 INFO:__main__:    return self._do_run(**kwargs)
2021-03-09 17:23:34,223.223 INFO:__main__:  File "/home/jcollin/workspace/ceph/build/../qa/tasks/vstart_runner.py", line 467, in _do_run
2021-03-09 17:23:34,224.224 INFO:__main__:    proc.wait()
2021-03-09 17:23:34,224.224 INFO:__main__:  File "/home/jcollin/workspace/ceph/build/../qa/tasks/vstart_runner.py", line 213, in wait
2021-03-09 17:23:34,224.224 INFO:__main__:    raise CommandFailedError(self.args, self.exitstatus)
2021-03-09 17:23:34,224.224 INFO:__main__:teuthology.exceptions.CommandFailedError: Command failed with status 22: ['./bin/ceph', 'fs', 'set', 'cephfs', 'allow_multimds_snaps', 'false']
```

File: src/mds/MDSMap.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (joscollin) [reply src/mds/MDSMap.cc]: Fixed this issue. Please check.

File: qa/tasks/cephfs/test_misc.py (hunk 9 above)
Comment (batrick) [reply qa/tasks/cephfs/test_misc.py:209]: There should be a message to `stderr`, what does it say?

File: src/mds/MDSMap.h [hunk 11]
```diff
@@ -231,6 +231,7 @@ class MDSMap {
   }
   void clear_multimds_snaps_allowed() { clear_flag(CEPH_MDSMAP_ALLOW_MULTIMDS_SNAPS); }
   bool allows_multimds_snaps() const { return test_flag(CEPH_MDSMAP_ALLOW_MULTIMDS_SNAPS); }
+  bool not_joinable() const { return test_flag(CEPH_MDSMAP_NOT_JOINABLE); }
```
Comment (batrick) [inline src/mds/MDSMap.h:234]: You might even do:

```suggestion
  bool joinable() const { return !test_flag(CEPH_MDSMAP_NOT_JOINABLE); }
```

File: qa/tasks/cephfs/test_misc.py (hunk 9 above)
Comment (joscollin) [reply qa/tasks/cephfs/test_misc.py]: I think Let's test if the `lsflags` print some default flag values too. Please see the latest test code.

**Comment by joscollin** [comment]
@batrick The review comments are fixed, please take a look.

**Comment by batrick** [comment]
jenkins test make check

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
https://pulpito.ceph.com/pdonnell-2021-04-01_00:45:34-fs-wip-pdonnell-testing-20210331.222326-distro-basic-smithi/

- https://tracker.ceph.com/issues/48772
    qa: pjd: not ok 9, 44, 80
- https://tracker.ceph.com/issues/50177
    osd: "stalled aio... buggy kernel or bad device?"
- https://tracker.ceph.com/issues/48771
    qa: iogen: workload fails to cause balancing
- https://tracker.ceph.com/issues/49845
    qa: failed umount in test_volumes
- https://tracker.ceph.com/issues/48773
    qa: scrub does not complete
- https://tracker.ceph.com/issues/48805
    mds: "cluster [WRN] Scrub error on inode 0x1000000039d (/client.0/tmp/blogbench-1.0/src/blogtest_in) see mds.a log and `damage ls` output for details"
- https://tracker.ceph.com/issues/50178
    qa: "TypeError: run() got an unexpected keyword argument 'shell'"
- https://tracker.ceph.com/issues/45434
    qa: test_full_fsync (tasks.cephfs.test_full.TestClusterFull) failed

**Review by batrick — CHANGES_REQUESTED** [review]
Jos, please add documentation to `doc/cephfs/administration.rst`.

**Comment by batrick** [comment]
Code is fine, we can merge after docs are done.

**Comment by joscollin** [comment]
jenkins render docs

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
jenkins test api

**Comment by joscollin** [comment]
jenkins test api

**Comment by joscollin** [comment]
jenkins retest this please

**Comment by joscollin** [comment]
@batrick Rebased, but still API tests fails with DashboardException.

**Comment by batrick** [comment]
`tasks.mgr.dashboard.helper._ValError: In `input['fs_map']['filesystems'][0]['mdsmap']`: unknown keys: {'flags_state'}`

@ceph/dashboard how can we fix this?

**Comment by epuertat** [comment]
> `tasks.mgr.dashboard.helper._ValError: In `input['fs_map']['filesystems'][0]['mdsmap']`: unknown keys: {'flags_state'}`
> 
> @ceph/dashboard how can we fix this?

@batrick in `qa/tasks/mgr/dashboard/test_health.py` the mds map is being validated, so a schema for this new section should be added between these lines:

https://github.com/ceph/ceph/blob/1b8ee16c13e3ab4d25de66ebfd12d8692658a6f3/qa/tasks/mgr/dashboard/test_health.py#L53-L54

How should these new flags be exposed in the Dashboard?

**Comment by batrick** [comment]
> How should these new flags be exposed in the Dashboard?

It's just an array of strings indicating the flags on the fs. It can be omitted or not.

**Comment by batrick** [comment]
@joscollin please make the corrections.

**Comment by joscollin** [comment]
jenkins retest this please

File: qa/tasks/mgr/dashboard/test_health.py [hunk 12]
```diff
@@ -51,6 +51,11 @@ class HealthTest(DashboardTestCase):
         'modified': str,
         'session_timeout': int,
         'flags': int,
+        'flags_state': JObj({}, allow_unknown=True),
+        'joinable': bool,
+        'allow_snaps': bool,
+        'allow_multimds_snaps': bool,
+        'allow_standby_replay': bool,
```
Comment (epuertat) [inline qa/tasks/mgr/dashboard/test_health.py:58]:
```suggestion
        'flags_state': JObj({
	          'joinable': bool,
  	          'allow_snaps': bool,
  	          'allow_multimds_snaps': bool,
  	          'allow_standby_replay': bool,
  	     }),
```

**Review by epuertat — CHANGES_REQUESTED** [review]

**Review by epuertat — APPROVED** [review]
Thanks a lot, @joscollin ! 👍🏼

**Comment by batrick** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2021-Apr-15
